abi.html

```diff
@@ -5827,6 +5957,298 @@ <h5><a href="#mangling.declaration-reference">5.1.6.2 References to declared ent
+the member. For example:
+
+<code><pre>
+struct A {
+    void foo();
+};
+template &lt;void (A::*)()&gt; void f();
+
+f&lt;&amp;A::foo&gt; // _Z1fIXadL_ZN1A3fooEvEEEvv
+</pre></code>
+
+<p>
+If the class type of the constant is not the declaring class of the
+member, the reference expression is wrapped in a special expression
+used only for this purpose, which encodes the <code>this</code>-adjustment
+that must be applied to a pointer to the class type of the constant
+in order to produce a pointer to the declaring class of the member.
+The offset is omitted if zero.  The notional type of this expression
+is a member pointer type using the class type of the constant and the
+declared type of the member.
+
+<pre><font color=blue><code>
+  &lt;expression&gt; ::= mc &lt;<i>class</i> <a href="#mangle.type">type</a>&gt; &lt;<a href="#mangle.expression">expression</a>&gt; [&lt;<i>offset</i> <a href="#mangle.number">number</a>&gt;] E
+</code></font></pre>
+
+<p>
+Finally, if precise typing is required and the notional type of the
+current reference expression is not the type of the constant, the
+reference expression is wrapped in a C-style cast to the constant
+type (that is, prefixed with <code class=mangle>cv &lt;type&gt;</code>).
+This can happen because <code>static_cast</code> may add qualifiers
+to the member type.
+
+<code><pre>
+struct A {
+  char x;
+  void foo();
+};
+struct B {
+  char y;
+};
+
+struct C : A, B {};
+template &lt;void (C::*)()&gt; void f();
+
+f&lt;&amp;A::foo&gt; // _Z1fIXmcM1CFvvEadL_ZN1A3fooEvE1EEEvv
+</pre></code>
+
+<p>
+The <code class=mangle>mc</code> mangling uses an offset instead of
+a base path to avoid exposing private details of the class into the ABI,
+such as the names of any intermediate base classes.
+
+<a name="mangling.member-initializer-sequence">
+<h5><a href="#mangling.member-initializer-sequence">5.1.6.5 Member initializer sequences</a></h5>
+</a>
+
+<p>
+The constant evaluation of an expression of class type assigns
+a constant value to each non-static data member of the class or,
+if the class is a union, to at most one non-static data member,
+called the active union member.  Under the standard, this exact
+structure uniquely determines a value, including whether a union
+has an active union member member and (if so) which one.  As
+different values can produce different template specializations,
+this structure must be faithfully represented in the mangling of
+a value as a template argument.
+
+<p>
+A value is converted to a flattened sequence of values by applying
+the following expansions until no values of array or non-union class
+type remain:
+<ul compact>
+  <li>A value of array type is replaced by its element values in
+    index order.
+  <li>A value of non-union class type is replaced by the values
+    of its base subobjects, in order of declaration, followed
+    by the values of its non-static data members, in order of
+    declaration.
+</ul>
```

Comment:
Clang and GCC don't do this flattening; this would be an ABI change for both: https://godbolt.org/z/nzWP13MMf ... but I think these manglings are new enough that we can afford to change them.

For the most part, this flattening seems good: it'll make manglings shorter in some cases, and avoids encoding details of the internal structure of the type (eg, empty base classes and additional layers of base class wrapping don't affect the mangling). But it also will cause problems for the tail-dropping below, for the motivating cases such as:

```
struct StringParam {
  char data[256];
  int length;
};
template<StringParam str> void f();
```
... where the rule used by Clang and GCC would mangle `f<{"foo", 3}>()` relatively compactly (`_Z1fIXtl11StringParamtlA256_cLc102ELc111ELc111EELi3EEEEvv`) but this flattening will mean that we can't do the tail-dropping any more and will encode 253 additional zero `Lc0E`s before the `Li3E`.

Perhaps we can keep the flattening of non-union class types and remove the flattening of arrays? Because we don't need precise typing, we can use `il` instead of `tl` here, rather than encoding the array type as Clang and GCC currently do.

Reply:
Not flattening arrays so that we can use tail optimization makes sense to me; I can write that up.